src/shell/output.c

```diff
@@ -63,8 +63,16 @@ enum {
     FLUX_OUTPUT_TYPE_FILE = 3,
 };
 
+struct shell_output_data {
+    void *buf;
+    int len;
+    int offset;
+};
+
```

Comment:
Is the "buffer watcher" not useful here?  E.g. `flux_buffer_write_watcher_create()`?

Reply:
main issue was buffer watcher has to have a buffer size set when it's created.  So I elected to go with the fd watcher.  

Thinking this through, could possible save data from `out->output` somewhere else when buffer is full.  Not sure if it's a net win code wise.

It did just occur me that I should probably add a completion reference for fd-watcher writes, since on exit, there's no way to know if there's still data for the fd watcher to keep on writing.    As a side effect, would clean up some of the `shell_output_destroy()` path.

Reply:
Ok, I'll have a closer look at this code then.  Are there tests that can make write(2) return EAGAIN/EWOULDBLOCK (seems like that would be hard to do if output is always a file).

Reply:
Nah, no tests for that specific path.  Hard to create a test.

BUT, I only checked for EAGAIN, totally forgot about EWOULDBLOCK.  Need to fix that up.

Reply:
oh wait, EWOULDBLOCK is only for sockets.  I had it right :-)

Reply:
Wait a second, the data is already "buffered" in the `output` JSON array, so we shouldn't need to buffer it again in the zlist.  It should be possible to get this done by just adding a fd watcher and an offset count for each of stderr/stdout:
* init offset to zero
* start fd watcher (POLLIN) on transition of output array from empty to non-empty
* on fd callback, decode "top" element of `output` and write(2) data+offset, len-offset
* if EAGAIN/EWOULDBLOCK, return (no-op)
* if full write, pop element, zero offset, and return (stop watcher if output is now empty)
* if partial write, add size to offset and keep writing until EAGAIN/EWOULDBLOCK

It may seem ugly for the case where you get EAGAIN/EWOULDBLOCK after a partial write to decode the same element multiple times.  However, it seems like that could be a future optimization if that code path is found to be "hot".  Its hard to know if it will be.

Also, the code to write to stdout/stderr for the standalone case could use this code path with fds set to STDOUT_FILENIO/STDERR_FILENO.

Reply:
> seems like that would be hard to do if output is always a file 

Yeah, I wonder how hard it would be to allow output to stdin (using `popen`) of a process instead of a file (for testing only). It would be nice, but probably a fair bit of code just for testing.

>  main issue was buffer watcher has to have a buffer size set when it's created. So I elected to go with the fd watcher.

I haven't looked at the code, but I guess it makes sense when you already have the  data buffered elsewhere (in the output json array here)

However, the underlying buffer *is* a cbuf, which can grow if necessary (if we exposed those hooks).

Reply:
> Also, the code to write to stdout/stderr for the standalone case could use this code path with fds set to STDOUT_FILENIO/STDERR_FILENO.

Which could be wired to a pipe or something to simulate EAGAIN/partial write handling as was done in librouter/test/sendfd.c.

Reply:
Great suggestion!

Reply:
Just saw @grondo's suggestion to use the write watcher.  Missed that earlier, so sorry for the redundant comment.

Reply:
No problem!

I really like generalizing the "file" output to "fd" output and switch terminal output to that same handler. However, I wonder if we are running out of time for that approach in this PR?

We could delay merging the PR and pursue this cleanup, or merge the (working) PR soon and do the cleanup as a follow-on PR. (I also kind of want to investigate moving the "output modes" into their own plugins, but haven't thought to hard about how that would work yet. But the output plugin could change regardless..)

I leave the approach up to @chu11 and @garlick

Reply:
That works for me.  @chu11, if that makes sense to you, then my vote would be to drop the "support nonblocking writes" commit, and then this is probably good to go in IMHO.

Reply:
> Wait a second, the data is already "buffered" in the output JSON array, so we shouldn't need to buffer it again in the zlist. 

So the main reason I copy to another list is b/c both stdout and stderr both go to the same output JSON array, but the streams could go to different locations, like file/kvs/terminal. or different files  One stream could be blocking and one not.

If I leave everything on the output array, then I have to iterate through the entire array everytime, even if one stream isn't blocked.  Perhaps this isn't the end of the world and not as big a deal as I originally thought.  Some logic has to be changed since I can't call `json_array_clear()` anymore, gotta remove elements from the output array as they are processed.

> if partial write, add size to offset and keep writing until EAGAIN/EWOULDBLOCK

(Note, I now see a mistake I did in `output.c`, so following sentence is in principle) One of the side reasons I did the list is b/c when you call `iodecode()` the `data` from the log entry is malloced when returned to the caller.  This decode+malloc has to be done.  It seemed like little cost to save off this already-malloced region of memory, rather than leave it on the `output` array to be re-malloced the next time your try to read from the array.

Reply:
Oh I missed that detail, sorry.  The streams mixed up in one array does make my proposal a bit trickier.

Reply:
> That works for me. @chu11, if that makes sense to you, then my vote would be to drop the "support nonblocking writes" commit, and then this is probably good to go in IMHO.

Ok, let me remove that commit and I'll re-push